sdk/storage/Azure.Storage.Common/tests/StorageWriteStreamTests.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Storage.Shared;
+using Moq;
+using NUnit.Framework;
+
+namespace Azure.Storage.Tests
+{
+    public class StorageWriteStreamTests
+    {
+        private static readonly string s_append = "Append";
+        private static readonly string s_flush = "Flush";
+
+        /// <summary>
+        /// This test uses a 1 KB buffer, and calls WriteAsync() 9 times.
+        /// We expect 2 Append calls, 1 Flush call, and 9 writes to the buffer.
+        /// In the current implementation, there will also be 2 writes to the buffer with count = 0.
+        /// </summary>
+        [Test]
+        public async Task BasicTest()
+        {
+            // Arrange
+            int bufferSize = Constants.KB;
+            int writeSize = 256;
+            int writeCount = 9;
+
+            Mock<PooledMemoryStream> mockBuffer = new Mock<PooledMemoryStream>(MockBehavior.Strict);
+            StorageWriteStreamImplementation stream = new StorageWriteStreamImplementation(
+                position: 0,
+                bufferSize: bufferSize,
+                progressHandler: null,
+                buffer: mockBuffer.Object);
+
+            mockBuffer.SetupSequence(r => r.WriteAsync(
+                It.IsAny<byte[]>(),
+                It.IsAny<int>(),
+                It.IsAny<int>(),
+                It.IsAny<CancellationToken>()))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(0))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(writeSize))
+                    .Returns(Task.FromResult(0))
+                    .Returns(Task.FromResult(writeSize));
+
+            mockBuffer.SetupSequence(r => r.Position)
+                .Returns(0)
+                .Returns(256)
+                .Returns(512)
+                .Returns(768)
+                .Returns(1024)
+                .Returns(1024)
+                .Returns(0)
+                .Returns(256)
+                .Returns(512)
+                .Returns(1024)
+                .Returns(1024);
```

Comment:
Does this imply that the `Position` of our write stream reflects where in the buffer we write the next piece of data? If I've written 2048 bytes to a stream then I expect `Position` to return 2048, not whatever index of the internal buffer we happen to be at.